bundles/org.openhab.binding.tesla/src/main/java/org/openhab/binding/tesla/internal/TeslaChannelSelectorProxy.java

```diff
@@ -162,40 +164,28 @@ public State getState(String s, TeslaChannelSelectorProxy proxy, Map<String, Str
         CHARGE_LIMIT_SOC_MIN("charge_limit_soc_min", "chargelimitminimum", PercentType.class, false),
         CHARGE_LIMIT_SOC_STD("charge_limit_soc_std", "chargelimitsocstandard", PercentType.class, false),
         CHARGE_PORT_LATCH("charge_port_latch", "chargeportlatch", StringType.class, false),
-        CHARGE_MILES_ADDED_IDEAL("charge_miles_added_ideal", "chargeidealmilesadded", DecimalType.class, false) {
+        CHARGE_MILES_ADDED_IDEAL("charge_miles_added_ideal", "chargeidealdistanceadded", DecimalType.class, false) {
             @Override
             public State getState(String s, TeslaChannelSelectorProxy proxy, Map<String, String> properties) {
                 State someState = super.getState(s);
-                double odo = ((DecimalType) someState).doubleValue();
-                if (properties.containsKey("distanceunits") && properties.get("distanceunits").equals("km/hr")) {
-                    return super.getState(String.valueOf(odo * 1.609344));
-                } else {
-                    return someState;
-                }
+                BigDecimal value = ((DecimalType) someState).toBigDecimal();
+                return new QuantityType<>(value, ImperialUnits.MILES_PER_HOUR);
             }
         },
-        CHARGE_MILES_ADDED_RANGE("charge_miles_added_rated", "chargeratedmilesadded", DecimalType.class, false) {
+        CHARGE_MILES_ADDED_RANGE("charge_miles_added_rated", "chargerateddistanceadded", DecimalType.class, false) {
             @Override
             public State getState(String s, TeslaChannelSelectorProxy proxy, Map<String, String> properties) {
                 State someState = super.getState(s);
-                double odo = ((DecimalType) someState).doubleValue();
-                if (properties.containsKey("distanceunits") && properties.get("distanceunits").equals("km/hr")) {
-                    return super.getState(String.valueOf(odo * 1.609344));
-                } else {
-                    return someState;
-                }
+                BigDecimal value = ((DecimalType) someState).toBigDecimal();
+                return new QuantityType<>(value, ImperialUnits.MILES_PER_HOUR);
             }
         },
         CHARGE_RATE("charge_rate", "chargerate", DecimalType.class, false) {
             @Override
             public State getState(String s, TeslaChannelSelectorProxy proxy, Map<String, String> properties) {
                 State someState = super.getState(s);
-                double odo = ((DecimalType) someState).doubleValue();
-                if (properties.containsKey("chargerateunits") && properties.get("chargerateunits").equals("km/hr")) {
-                    return super.getState(String.valueOf(odo * 1.609344));
```

Comment:
Is it correct that this was the same as line 475 but ends up differently? I see no reason that they should not be the same still, what do you think?

Reply:
Not sure what you exactly mean here.
https://github.com/openhab/openhab2-addons/pull/6342/files#diff-a5056d3cb3275cb84a62e77fdd7ad0e2R465-R466 is the same, the only difference is the unit (here it is `mph`, while below it is `miles`).

Reply:
So in this case you rewrote:

```java
double odo = ((DecimalType) someState).doubleValue();
if (properties.containsKey("chargerateunits") && properties.get("chargerateunits").equals("km/hr")) {
    return super.getState(String.valueOf(odo * 1.609344));
} else {
    return someState;
}
```

To: 

```java
BigDecimal value = ((DecimalType) someState).toBigDecimal();
return new QuantityType<>(value, ImperialUnits.MILES_PER_HOUR);
```

while around line 321 (maybe not around 475...):

```java
double odo = ((DecimalType) someState).doubleValue();
if (properties.containsKey("distanceunits") && properties.get("distanceunits").equals("km/hr")) {
    return super.getState(String.valueOf(odo * 1.609344));
} else {
    return someState;
}
```

Is converted to:

```java
BigDecimal value = ((DecimalType) someState).toBigDecimal();
if (properties.containsKey("distanceunits") && properties.get("distanceunits").equals("km/hr")) {
    return new QuantityType<>(value, MetricPrefix.KILO(SIUnits.METRE));
} else {
    return new QuantityType<>(value, ImperialUnits.MILE);
}
```

I would think we should always be able to depend on the internal conversion system.  But at least I expect the same code to be rewritten to the same code.

Reply:
Now that you have pointed me to the right place (line 321 instead of 475), I have to fully agree with you 😆. Fixed it now!